SpherePacking/MagicFunction/b/psi.lean

```diff
@@ -51,65 +51,449 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+    exact d_nonzero
+  rw [resolve']
+  rw [sub_neg_eq_add, ← add_div, mul_assoc, div_mul]
+  rw [div_self, div_one, left_distrib, add_comm, left_distrib]
+  exact pow_ne_zero 2 (z_plus_one_nonzero z)
+  exact pow_ne_zero 2 (z_plus_one_nonzero z)
+  exact pow_ne_zero 2 (z_plus_one_nonzero z)
+-- proof of ψS_slash_ST complete
+
+-- The following lemma first was
+-- lemma ψS_slash_T : ψS ∣[-2] T = ψS := by
+-- but it turns out there needs to be a -
+lemma ψS_slash_T : ψS ∣[-2] T = -ψS := by
+  ext z
+  rw [ψS_eq', slashT']
+  simp only [Pi.mul_apply, Pi.add_apply, Pi.div_apply,
+    Pi.pow_apply,  Pi.sub_apply, smul_add]
+  rw [← slashT z ⇑H₂_MF, ← slashT z ⇑H₃_MF, ← slashT z ⇑H₄_MF]
+  have hh2 : (H₂_MF : ℍ → ℂ) = H₂ := by exact rfl
+  have hh3 : (H₃_MF : ℍ → ℂ) = H₃ := by exact rfl
+  have hh4 : (H₄_MF : ℍ → ℂ) = H₄ := by exact rfl
+  rw [hh2 , hh3, hh4]
+  rw [H₂_T_action, H₃_T_action, H₄_T_action]
+  field_simp
+  rw [sub_eq_add_neg, add_comm,  ← sub_neg_eq_add, ← neg_sub', mul_neg, add_comm, ← sub_eq_add_neg,
+      add_comm (H₃ z) _ ]
+-- proof of ψS_slash_T complete
+
+lemma ψT_slash_S : ψT ∣[-2] S = -ψT := by
+  ext z
+  rw [ψT_eq, slashS']
+  simp only [Pi.mul_apply, Pi.ofNat_apply, Pi.add_apply, Pi.div_apply,
+    Pi.pow_apply, Pi.neg_apply]
+  rw [slashS'' z ⇑H₂_MF, slashS'' z ⇑H₃_MF, slashS'' z ⇑H₄_MF]
+  have hh2 : (H₂_MF : ℍ → ℂ) = H₂ := by exact rfl
+  have hh3 : (H₃_MF : ℍ → ℂ) = H₃ := by exact rfl
+  have hh4 : (H₄_MF : ℍ → ℂ) = H₄ := by exact rfl
+  rw [hh2 , hh3, hh4]
+  rw [H₂_S_action, H₃_S_action, H₄_S_action]
+  field_simp
+  have resolve (h2 h3 h4 d : ℂ) (d_nonzero : d ≠ 0) : (- (h4 * d) + - (h3 * d)) / (h2 * d) ^ 2 =
+                                                       - ((h4 + h3) / (h2 ^ 2) / d ) := by
+    rw [← neg_add, ← neg_div', ← right_distrib]
+    rw [pow_two, mul_comm h2 d, mul_assoc d h2 (d * h2), ← div_div, ← mul_div_assoc']
+    rw [div_self, mul_one, mul_comm d h2, ← mul_assoc, ← div_div, ← pow_two]
+    exact d_nonzero
+  rw [resolve, resolve]
+  rw [neg_div', neg_div', neg_div', ← add_div, mul_div_assoc',  div_mul]
+  rw [div_self, div_one, neg_div, ← sub_eq_add_neg, ← neg_add', mul_neg, add_comm, add_comm (H₄ z) _, add_comm (H₃ z) (H₂  z)]
+  exact pow_ne_zero 2 (z_nonzero z)
+  exact pow_ne_zero 2 (z_nonzero z)
+  exact pow_ne_zero 2 (z_nonzero z)
+-- proof of ψT_slash_S complete
+
+lemma ψI_slash_TS : ψI ∣[-2] (T * S) = -ψT := by
+  ext z
+  rw [slash_mul]
+  have def_ψT : ψT = ψI ∣[-2] (T) := rfl
+  rw [← def_ψT, ψT_slash_S]
+
+
+lemma ψS_slash_STS : ψS ∣[-2] (S * T * S) = -ψT := by
+  ext z
+  rw [slash_mul, slash_mul]
+  rw [ψS_slash_S, ]
+  have def_ψT : ψT = ψI ∣[-2] (T) := rfl
+  rw [← def_ψT, ψT_slash_S]
+
+-- original statement of this lemm contained a sign error:
+-- lemma ψS_slash_TSTS : ψS ∣[-2] (T * S * T * S) = -ψT
+lemma ψS_slash_TSTS : ψS ∣[-2] (T * S * T * S) = ψT := by
+  ext z
+  rw [slash_mul, slash_mul, slash_mul]
+  rw [ψS_slash_T, neg_slash, ψS_slash_S, neg_slash]
```

Comment:
Again, the `rw` lines here can be merged. It's considered better style to have all rewrite statements in one line. So this line would need to look like
```
rw [slash_mul, slash_mul, slash_mul, ψS_slash_T, neg_slash, ψS_slash_S, neg_slash]
```